services/ymax-planner/src/support.ts

```diff
@@ -102,3 +102,88 @@ export const createEVMContext = async ({
     usdcAddresses: usdcAddresses[clusterName],
   };
 };
+
+const findBlockByTimestamp = async (
+  provider: JsonRpcProvider,
+  targetMs: number,
+) => {
+  const target = Math.floor(targetMs / 1000);
+  let latest = await provider.getBlockNumber();
+  let earliest = 0;
+
+  while (earliest <= latest) {
+    const mid = Math.floor((earliest + latest) / 2);
+    const block = await provider.getBlock(mid);
+    if (!block) break;
+
+    if (block.timestamp === target) return mid;
+    if (block.timestamp < target) earliest = mid + 1;
+    else latest = mid - 1;
+  }
+  // latest is now the greatest block with timestamp <= target
+  return latest;
+};
+
+export const buildTimeWindow = async (
+  provider: JsonRpcProvider,
+  publishTimeMs: number,
+) => {
+  const fromBlock = await findBlockByTimestamp(provider, publishTimeMs);
```

Comment:
Binary search against external APIs seems like a sufficiently common need (and with sufficiently sharp edges) that we should just have a helper in package "internal".
```suggestion
  const posixSeconds = Math.floor(publishTimeMs / 1000);
  // <T, U=T>binarySearch(
  //   start: T,
  //   end: T,
  //   acceptancePredicate: (x: U) => Promise<boolean> | boolean,
  //   transform?: (x: T) => Promise<U> | U
  // ): T
  // assumes a transition from acceptance to rejection somewhere in [start, end]
  // and returns the greatest accepted T
  const startBlockNumber = await binarySearch(
    0,
    await provider.getBlockNumber(),
    block => block.timestamp <= posixSeconds,
    blockNumber => provider.getBlock(blockNumber),
  );
```